Added Asset Ranking and Sorting (Plus refactored)

ziadoz opened the pull request:
- Refactored out Assets into their own class `Pangaea\Asset\Asset` that implements `Renderable`.
- Added a new helper `Xml::renderRenderables()` that will render out an array of renderables into a string.
- Updated `Item` to use new `Asset` class without changing method signature.
- Updated assets to be naturally sorted and then ranked.

gabriel403 commented:
Should asset work out type depending on rank?

ziadoz commented:
@gabriel403 I've fixed a few bugs in the ranking. The sorting was broken (because `natsort()` doesn't change the keys, just reorders the elements). I've also made it mark the first item in the original list of URLs passed as `PRIMARY`, which is consistent with its previous behaviour.

gabriel403 commented on src/Asset/Asset.php:
```diff
@@ -0,0 +1,177 @@
+     * @param int|null    $rank
+     * @param array|null  $attributes
+     */
+    public function __construct($url, $type = null, $rank = null, array $attributes = null)
```
is there any circumstances when any of these would actually be missing? Looking at the implementation below I think it might make more sense to remove the defaults?

ziadoz replied:
@gabriel403 Well, we'll always populate every parameter, but in terms of the actual XML/XSDs only the URL and type (which I've updated now to be required) are actually mandatory.

gabriel403 commented:
:+1:

gabriel403 commented on src/Xml.php:
```diff
@@ -110,4 +110,24 @@ private static function errorSummary()
+     */
+    public static function renderRenderables(array $renderables)
+    {
+        if (count($renderables) === 0) {
```
do you need this? I think the array_reduce with an empty array would product this anyway?

ziadoz commented:
@gabriel403 Remove the count check, fixed a bug in the XML function, and added unit tests for the new render helper method (using Mockery for the mocks).

gabriel403 commented:
:+1:
